src/tools/Avalonia.Generators/NameGenerator/AvaloniaNameIncrementalGenerator.cs

```diff
@@ -0,0 +1,204 @@
+                                {
+                                    continue;
+                                }
+
+                                resolvedNames.Add(nameResolver
+                                    .ResolveName(clrType, xmlName.Name, xmlName.FieldModifier));
+                            }
+                            catch (Exception ex)
+                            {
+                                diagnostics.Add(GeneratorExtensions.NameGeneratorUnhandledError(ex));
+                            }
+                        }
+
+                        view = new ResolvedView(xmlView, type.IsAvaloniaWindow(), resolvedNames.ToImmutableArray());
+                    }
+                }
+
+                return new ResolvedClassInfo(view, hasDevToolsReference, diagnostics.ToImmutableArray());
+            })
+            .WithTrackingName(TrackingNames.ResolvedNamesProvider);
+
+        context.RegisterSourceOutput(resolvedNames.Combine(options), static (context, pair) =>
+        {
+            var (info, options) = pair;
+
+            foreach (var diagnostic in info!.Diagnostics)
+            {
+                context.Report(diagnostic);
+            }
+
+            if (info.View is { } view && options.AvaloniaNameGeneratorFilterByNamespace.Matches(view.Namespace))
+            {
+                ICodeGenerator codeGenerator = options.AvaloniaNameGeneratorBehavior switch
+                {
+                    Behavior.OnlyProperties => new OnlyPropertiesCodeGenerator(
+                        options.AvaloniaNameGeneratorClassFieldModifier),
+                    Behavior.InitializeComponent => new InitializeComponentCodeGenerator(
+                        options.AvaloniaNameGeneratorAttachDevTools && info.CanAttachDevTools && view.IsWindow,
+                        options.AvaloniaNameGeneratorClassFieldModifier),
+                    _ => throw new ArgumentOutOfRangeException()
+                };
+                var fileName = options.AvaloniaNameGeneratorViewFileNamingStrategy switch
+                {
+                    ViewFileNamingStrategy.ClassName => $"{view.ClassName}.g.cs",
+                    ViewFileNamingStrategy.NamespaceAndClassName => $"{view.Namespace}.{view.ClassName}.g.cs",
+                    _ => throw new ArgumentOutOfRangeException(
+                        nameof(ViewFileNamingStrategy), options.AvaloniaNameGeneratorViewFileNamingStrategy,
+                        "Unknown naming strategy!")
+                };
+
+                var generatedPartialClass = codeGenerator.GenerateCode(
+                    info.View.ClassName,
+                    info.View.Namespace,
+                    info.View.Names);
+
+                context.AddSource(fileName, generatedPartialClass);
+            }
+        });
+    }
+
+    internal record XmlClassInfo(
+        ResolvedXmlView? XmlView,
+        ImmutableArray<DiagnosticDescriptor> Diagnostics);
```

Comment:
`ImmutableArray` equality is comparing the underlying array instead of its items so caching won't work. This needs a wrapper type.

Reply:
I have seen immutable array used with source generators. And roslyn has ImmutableArrayComparer implemented... It's not a clear answer though, as I don't know if this comparer is actually used here

Reply:
Replaced with EquatableList